### ISSUE TITLE
Lint against imports that are dead code

### ISSUE THREAD

**Issue by CryZe** [rust-lang/rust#80443]
I'm very often seeing people write imports such as this one:
`use regex;`

Such an import is almost always dead code, because ever since Rust 2018 all external crates are already in scope. In fact pretty much any import that doesn't contain at least one of `pub`, `as ...` or `::` is very likely dead code. Rust does not currently emit any warning for such an import.

**Comment by jonas-schievink** [comment]
`use regex;` on the playground does emit a warning:

```
warning: unused import: `regex`
 --> src/lib.rs:1:5
  |
1 | use regex;
  |     ^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
```

**Comment by CryZe** [comment]
That's just the default warning of an unused import, not the specific case I'm talking about. Here's a proper reproduction:

```rust
use regex;

fn main() {
    regex::Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();
}
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=186c293aa81cfe34eeeea813e4f3876f)

**Comment by ThePuzzlemaker** [comment]
I think I'll take a shot at this. @rustbot claim

**Comment by ThePuzzlemaker** [comment]
Should this be implemented as a new lint or just as an instance of `dead_code`? I personally think it'd be better as its own lint (something like `useless_import`) as dead code is described in `rustc_lint_defs` as "detect unused, unexported items"

**Comment by petrochenkov** [comment]
We already have a lint for redundant imports (https://github.com/rust-lang/rust/pull/58805), it's a part of `unused_imports`, but it conservatively doesn't fire on imports in `mod` items.
The reason is that such imports can be used by module-relative paths like `super::regex` from a child module or `crate::regex`/`self::regex` from the current module, but we don't currently track such uses. See the linked PR and issue for more details.

**Comment by petrochenkov** [comment]
This is also a duplicate of https://github.com/rust-lang/rust/issues/61640.

**Comment by rylev** [comment]
Triage: going to close this in favor of #61640.
